// dhcp/index.js
var util = require('util'),
    dhcp = require('./dhcp'),
    dgram = require('dgram');

// constructor
var Server = function() {
  dgram.Socket.apply(this, arguments);
  
  // enable broadcasts
  this.setBroadcast(true);
  
  var dummyCallback = function () { return {}; }
  this.discoverCallback = dummyCallback;
  this.requestCallback = dummyCallback;
  this.declineCallback = dummyCallback;
  this.releaseCallback = dummyCallback;
  this.informCallback = dummyCallback;
  
  var self = this;
  // handle incoming packets
  self.on('message', function(buffer, remote) {
    
    var packet = dhcp.Packet.fromBuffer(buffer);
    if (packet.op == 1) {
      
      var type_id = packet.options[53] || 0,
          type_descr = dhcp.MessageTypes[type_id];
          
      util.log('Got '+type_descr+' from '+remote.address+':'+remote.port+' ('+packet.chaddr+') ' +
        'with packet length of '+buffer.length+' bytes');
      
      switch (parseInt(type_id)) {
        case 1: self._handle('discover', packet); break; // discover, send offer
        case 3: self._handle('request', packet); break; // request, send ack
        case 4: self._handle('decline', packet); break; // decline, notify admin
        case 7: self._handle('release', packet); break; // release, release ip address, keep record
        case 8: self._handle('inform', packet); break; // inform, send ack to ciaddr, without yiaddr/lease time
      }
      
    } else {
      console.log('  Unsupported message format');
    }
    
  });
  
  // process.on('uncaughtException', function(err) {
  //   require('util').debug('Uncaught Exception: ' + err);
  // });

};
util.inherits(Server, dgram.Socket);

// sends a packet via Socket.send() to ip an emits eventType
Server.prototype._send = function(eventType, ip, packet) {
  this.emit(eventType, packet);
  var self = this, buffer = dhcp.Packet.toBuffer(packet);
  this.send(buffer, 0, buffer.length, 68, ip, function(err, bytes) {
    if (err) 
      self.emit(eventType+'Error', err, packet);
    else 
      self.emit(eventType+'Sent', bytes, packet);
  });
}

// call message callbacks to retreive params
// build a dgram packet and send it to the client
Server.prototype._handle = function(messageType, packet) {
  this.emit(messageType, packet, packet.options[50] || null);
};

Server.prototype.offer = function(packet, params) {
  if (params) {
    packet.yiaddr = params.yiaddr;
    packet.siaddr = params.siaddr;
  
    // set options
    packet.options = params.options;
  }
  
  // ensure response message type
  packet.op = 2;
  packet.options[53] = 2;
  this._send('offer', '255.255.255.255', packet);
};
Server.prototype.ack = function(packet, params) {
  if (params) {
    packet.yiaddr = params.yiaddr;
    packet.siaddr = params.siaddr;
  
    // set options
    packet.options = params.options;
  }
  
  // ensure response message type
  packet.op = 2;
  packet.options[53] = 5;
  this._send('ack', '255.255.255.255', packet);
};
Server.prototype.nak = function(packet, params) {
  packet.op = 2;
  packet.options[53] = 6;
  this._send('nak', packet.ciaddr, packet);
}
Server.prototype.inform = function(packet, params) {
  if (params) {
    packet.yiaddr = params.yiaddr;
    packet.siaddr = params.siaddr;
  
    // set options
    packet.options = params.options;
  }
  
  // ensure response message type
  packet.op = 2;
  packet.options[53] = 5;
  this._send('inform', packet.ciaddr, packet);
};

module.exports = {
  createServer: function(type) {
    return new Server(type);
  },
  Server: Server,
  Packet: dhcp.Packet
}
